include/graphblas/algorithms/gmres.hpp

```diff
@@ -0,0 +1,428 @@
+		 *
+		 * The GMRES algorithm requires three workspace buffers with capacity \f$ n \f$:
+		 *
+		 * @param[in,out] r              A temporary vector of the same size as \a x.
+		 * @param[in,out] u              A temporary vector of the same size as \a x.
+		 * @param[in,out] temp           A temporary vector of the same size as \a x.
+		 *
+		 * Finally, the algebraic structures over which the GMRES is executed are given:
+		 *
+		 * @param[in]     ring           The semiring under which to perform the GMRES.
+		 * @param[in]     minus          The inverse of the additive operator of
+		 *                               \a ring.
+		 * @param[in]     divide         The inverse of the multiplicative operator
+		 *                               of \a ring.
+		 *
+		 * This algorithm may return one of the following error codes:
+		 *
+		 * @returns #grb::SUCCESS  When the algorithm has converged to a solution
+		 *                         within the given \a max_iterations and \a tol.
+		 * @returns #grb::FAILED   When the algorithm did not converge within the
+		 *                         given \a max_iterations.
+		 * @returns #grb::ILLEGAL  When \a A is not square.
+		 * @returns #grb::MISMATCH When \a x or \a b does not match the size of \a A.
+		 * @returns #grb::ILLEGAL  When \a x does not have capacity \f$ n \f$.
+		 * @returns #grb::ILLEGAL  When at least one of the workspace vectors does not
+		 *                         have capacity \f$ n \f$.
+		 * @returns #grb::ILLEGAL  If \a tol is not strictly positive.
+		 * @returns #grb::PANIC    If an unrecoverable error has been encountered. The
+		 *                         output as well as the state of ALP/GraphBLAS is
+		 *                         undefined.
+		 *
+		 * \par Performance semantics
+		 *
+		 *   -# This function does not allocate nor free dynamic memory, nor shall it
+		 *      make any system calls.
+		 *
+		 * For performance semantics regarding work, inter-process data movement,
+		 * intra-process data movement, synchronisations, and memory use, please see
+		 * the specification of the ALP primitives this function relies on. These
+		 * performance semantics, with the exception of getters such as #grb::nnz, are
+		 * specific to the backend selected during compilation.
+		 */
+		template<
+			Descriptor descr = descriptors::no_operation,
+			typename IOType,
+			typename ResidualType,
+			typename NonzeroType,
+			typename InputType,
+			// typename PrecondmatrixType = nullptr_t,
+			class Ring = Semiring<
+				grb::operators::add< IOType >, grb::operators::mul< IOType >,
+				grb::identities::zero, grb::identities::one
+			>,
+			class Minus = operators::subtract< IOType >,
+			class Divide = operators::divide< IOType >
+
+		>
+		grb::RC gmres(
+			grb::Vector< IOType > &x,
+			const grb::Matrix< NonzeroType > &A,
+			const grb::Vector< InputType > &b,
+			std::vector< std::vector< NonzeroType > > &Hmatrix,
```

Comment:
Would recommend `double * const Hmatrix, and stipulate somewhere that its size is `k*k` with `k` the minimum of `n_restart` and `size(A)`, and stipulating that it's a row-major orientation. That should keep this output compatible with most other packages right, and seems like the best option until such time we have ALP/Dense integrated with ALP/GraphBLAS?

Reply:
I tried not to use c-arrays directly in the c++ code, @anyzelman perhaps you prefer to use `std::vector<>` instead of `std::vector< std::vector<> >`, or you prefer to use c-array directly? I can also leave it as a template in .hpp..

Reply:
A single (non-nested) `std::vector` is also OK for me